src/shell/braces.zig

```diff
@@ -0,0 +1,805 @@
+fn expandFlat(
+    tokens: []const Token,
+    expansion_table: []const ExpansionVariant,
+    out: []std.ArrayList(u8),
+    out_key: u16,
+    out_key_counter: *u16,
+    depth_: u8,
+    start: usize,
+    end: usize,
+) !void {
+    log("expandFlat [{d}, {d}]", .{ start, end });
+    if (start >= tokens.len or end > tokens.len) return;
+
+    var depth = depth_;
+    for (tokens[start..end], start..) |atom, j| {
+        _ = j;
+        switch (atom) {
+            .text => |txt| {
+                try out[out_key].appendSlice(txt.slice());
+            },
+            .close => {
+                depth -= 1;
+            },
+            .open => |expansion_variants| {
+                depth += 1;
+                if (bun.Environment.allow_assert) {
+                    std.debug.assert(expansion_variants.end - expansion_variants.idx >= 1);
+                }
+
+                var variants = expansion_table[expansion_variants.idx..expansion_variants.end];
+                const skip_over_idx = variants[variants.len - 1].end;
+
+                const starting_len = out[out_key].items.len;
+                for (variants[0..], 0..) |*variant, i| {
+                    const new_key = if (i == 0) out_key else brk: {
+                        const new_key = out_key_counter.*;
+                        try out[new_key].appendSlice(out[out_key].items[0..starting_len]);
+                        out_key_counter.* += 1;
+                        break :brk new_key;
+                    };
+                    try expandFlat(tokens, expansion_table, out, new_key, out_key_counter, depth, variant.start, variant.end);
+                    try expandFlat(tokens, expansion_table, out, new_key, out_key_counter, depth, skip_over_idx, end);
+                }
+                return;
+            },
+            else => {},
+        }
+    }
+}
+
+// pub fn expandNested()
+
+pub fn calculateVariantsAmount(tokens: []const Token) u32 {
+    var brace_count: u32 = 0;
+    var count: u32 = 0;
+    for (tokens) |tok| {
+        switch (tok) {
+            .comma => count += 1,
+            .open => brace_count += 1,
+            .close => {
+                if (brace_count == 1) {
+                    count += 1;
+                }
+                brace_count -= 1;
+            },
+            else => {},
+        }
+    }
+    return count;
+}
+
+const ParserError = error{
+    UnexpectedToken,
+};
+
+pub const Parser = struct {
+    current: usize = 0,
+    tokens: []const Token,
+    alloc: Allocator,
+    errors: std.ArrayList(Error),
+
+    // FIXME error location
+    const Error = struct { msg: []const u8 };
+
+    pub fn init(tokens: []const Token, alloc: Allocator) Parser {
+        return .{
+            .tokens = tokens,
+            .alloc = alloc,
+            .errors = std.ArrayList(Error).init(alloc),
+        };
+    }
+
+    pub fn parse(self: *Parser) !AST.Group {
+        var group_alloc_ = std.heap.stackFallback(@sizeOf(AST.Atom), self.alloc);
```

Comment:
This memory is invalid. We cannot use stackFallback and return values from the array without cloning them to the heap.